buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs

```
@@ -0,0 +1,123 @@
+    public async Task<string> GetAvailableTestRunner([FromBody] GetAvailableTestRunnerInput input)
+    {
+        if (string.IsNullOrEmpty(input.Roles))
+        {
+            throw new ArgumentNullException(nameof(input.Roles));
+        }
+        if (string.IsNullOrEmpty(input.ProjectName))
+        {
+            throw new ArgumentNullException(nameof(input.ProjectName));
+        }
+        if (string.IsNullOrEmpty(input.Branch))
+        {
+            throw new ArgumentNullException(nameof(input.Branch));
+        }
+
+        var roles = input.Roles.Split(",").Select(x => x.Trim()).ToList();
+
+        if (!roles.Any())
+        {
+            throw new ArgumentNullException(nameof(input.Roles));
+        }
+
+        foreach (var role in roles)
+        {
+            var assumeRoleResponse =
+                await new AmazonSecurityTokenServiceClient().AssumeRoleAsync(
+                    new AssumeRoleRequest
+                    {
+                        RoleArn = role,
+                        RoleSessionName = "DeployToolTestRunner"
+                    }
+                );
+
+            var stsClient = new AmazonSecurityTokenServiceClient(assumeRoleResponse.Credentials);
+            var callerIdentity = await stsClient.GetCallerIdentityAsync(new GetCallerIdentityRequest());
+
+            var codeBuildClient = new AmazonCodeBuildClient(assumeRoleResponse.Credentials);
+
+            var batchGetProjectsResponse =
+                await codeBuildClient.BatchGetProjectsAsync(
+                    new BatchGetProjectsRequest
+                    {
+                        Names = new List<string> { input.ProjectName }
+                    }
+                );
+
+            if (!batchGetProjectsResponse.Projects.Any())
+            {
+                throw new Exception($"Could not find any project with the name '{input.ProjectName}' in account '{callerIdentity.Account}'.");
```

If the project doesn't exist in a particular role should we just skip the role. I'm imagining when this function might be used for a collection of repositories and maybe we haven't created the appropriate codebuild job in all of our test accounts.

---

I threw an exception purposefully to indicate that there is an issue with the infrastructure setup instead of failing silently and us having a hard time tracking why the job was not triggered. This Lambda function is reading IAM roles from the GitHub repo secrets. If the secrets have the IAM role, then I think it's safe to assume that the infrastructure was properly setup, since that IAM role is automatically generated by the CF template that I am using to create the Codebuild job as well.